feat(order-manager): add an order manager to place and track orders

DCPRevere opened the pull request:
The Order Manager will track the status of orders on the DEX exchange.

It accepts orders from the strategy manager (and possible even manually placed orders).

And put them on the exchange using the DEX adapter. The DEX adapter alerts the order manager to status changes for the orders. This will then trigger it to place orders on the CEX.

DCPRevere committed:
feat: add an order manager to track order status

DCPRevere committed:
feat: add dex adapter empty class

DCPRevere committed:
fix: add header guard

DCPRevere committed:
feat: add order status and change structs

DCPRevere committed:
feat: add dex and cex

DCPRevere committed:
feat: add change_order_status method

DCPRevere committed:
feat: add placeholder tests for the order manager

DCPRevere committed:
feat: add methods for getting order sets and the full registry

DCPRevere committed:
test: add order_manage::place_order test

Milerius commented on src/dex/dex.cpp:
```diff
@@ -0,0 +1,22 @@
+
+namespace antara
+{
+
```
The pull request seem's imcomplete with empty files

DCPRevere replied:
As the scope of this PR is just the order manager, I may leave the CEX and DEX classes empty. Then, a later PR can add their functionality.

Milerius replied:
I would prefer not merge empty file and code that we are not using if it's possible it's will decrease our coverage once @ca333 will enable it

DCPRevere replied:
I think that this is an inevitability sometimes; otherwise, I would need to write strictly in a bottom-up fashion. This way, when someone comes along to write the DEX class, they know what what it's called and which methods need to be implemented.

DCPRevere replied:
The alternatives would be to either write the CEX and DEX first, or have half-empty implementations in the order manager functions.

DCPRevere replied:
I've changed them to throw `not_implemented` exceptions.

Milerius commented on src/order_manager/order.manager.cpp:
```diff
@@ -0,0 +1,59 @@
+    {
+        // for manually placing orders
+        // the strat manager will use place_orders
+        // registry_order_sets.emplace
```
Need's to be completed

DCPRevere commented:
Sorry, I should have said that this is incomplete.

DCPRevere committed:
fix: add const to ref argument

DCPRevere committed:
style: wrap long lines

DCPRevere committed:
feat: add orders and executions

Milerius commented on src/utils/mmbot_strong_types.hpp:
```diff
@@ -64,6 +64,12 @@ namespace antara
             st::addable_with<double>
     >;
 
+    using st_maker = st::type<
```
for boolean you are not forced to use strong type.

a simple `using maker = bool;` is ok

Milerius replied:
That's really strange to have a st::arithmetic on bool

DCPRevere replied:
I'm surprised it even compiled! Fixed.

DCPRevere committed:
feat: add order ids

DCPRevere committed:
feat: add order id hash function

DCPRevere committed:
feat: add pair helper function

DCPRevere committed:
feat: create the order class

feat: order manager holds orders by their id
feat: add place method to the dex class

DCPRevere committed:
feat: create order builder class

DCPRevere committed:
test: update tests to work with orders

Milerius commented:
I think you need to sync with master branch (several fix)

DCPRevere committed:
Merge remote-tracking branch 'origin' into add-order-manager

DCPRevere committed:
fix: incorrect usage of builder pattern

DCPRevere committed:
fix: use .value comparison for strong strings (for Windows)

DCPRevere committed:
fix: compare prices with .value (for Windows)

Milerius commented on src/cex/cex.cpp:
```diff
@@ -0,0 +1,25 @@
+{
+    void antara::cex::place_order(const orders::order_level &ol)
+    {
+        return;
```
This function need to be coded

DCPRevere replied:
This function is outside the scope of the PR and won't be completed.

Milerius commented on src/dex/dex.cpp:
```diff
@@ -0,0 +1,25 @@
+{
+    st_order_id dex::place(const orders::order &o)
+    {
+        return st_order_id{"A"};
```
This is spurious

DCPRevere replied:
This function is outside the scope of the PR and won't be completed.

Milerius commented on src/order_manager/order.manager.cpp:
```diff
@@ -0,0 +1,76 @@
+        // for manually placing orders
+        // the strat manager will use place_orders
+        // registry_order_sets.emplace
+        return true;
```
This function still need to be completed

Milerius commented on src/orders/orders.cpp:
```diff
@@ -20,10 +20,101 @@
 {
+    order::order(st_order_id &id, antara::pair &pair, st_price &price,
+                st_quantity &quantity, st_quantity &filled, antara::side &side,
+                order_status &status)
```
This need to be initialized with a braces list initialization https://en.cppreference.com/w/cpp/language/initializer_list

Milerius commented on src/orders/orders.hpp:
```diff
@@ -38,4 +38,73 @@ namespace antara::orders
+    public:
+        st_order_id id;
+
+        antara::pair pair;
```
all this field should be private i think

DCPRevere committed:
style: change st_maker to be a simple bool

DCPRevere committed:
style: use a braced-list style constr for order

DCPRevere committed:
feat: add execution id

DCPRevere committed:
feat: expand cex and dex methods

DCPRevere committed:
feat: rewrite the order_manager

style: move pair::of defn

DCPRevere committed:
feat: order_manager --> 2.0

Milerius commented on src/dex/dex.cpp:
```diff
@@ -0,0 +1,63 @@
+
+    orders::order dex::get_order_status(const st_order_id &id)
+    {
+        antara::pair pair = pair.of("B", "C");
```
Its really spurious here to use litteral like "A" "B" or "C", it's should be like that only in unit tests

Milerius commented on src/order_manager/order.manager.cpp:
```diff
@@ -0,0 +1,107 @@
+        }
+
+        // when an order is finished, remove it's executions
+        for (const auto& [id, order] : orders_) {
```
this should be a place for erase if idiom: https://en.wikipedia.org/wiki/Erase–remove_idiom

DCPRevere replied:
I thought the same, but I also need to remove executions here.
```
        for (const auto& [id, order] : orders_) {
            if (order.finished()) {
                std::vector<st_execution_id> ex_ids = order.execution_ids;
                for (const auto& id : ex_ids) {
                    executions_.erase(id.value());
                }
                orders_.erase(order.id.value());
            }
        }
```
And I'm thinking the idiom wouldn't work in this case.

DCPRevere replied:
Please tell me if you can think of a nicer way to write this.

Milerius commented on src/order_manager/order.manager.cpp:
```diff
@@ -0,0 +1,107 @@
+
+        auto all_executions = dex_.get_executions(order_ids);
+
+        for (const auto& ex : dex_.get_recent_executions()) {
```
this an std::copy: https://en.cppreference.com/w/cpp/algorithm/copy

Milerius commented on src/order_manager/order.manager.cpp:
```diff
@@ -0,0 +1,107 @@
+        }
+
+        // get all their executions
+        auto order_ids = std::vector<st_order_id>();
```
this an std::copy: https://en.cppreference.com/w/cpp/algorithm/copy

Milerius commented on src/order_manager/order.manager.cpp:
```diff
@@ -0,0 +1,107 @@
+
+        // add new orders
+        for (const auto& o : dex_.get_live_orders()) {
+            orders_.emplace(o.id.value(), o);
```
this an std::copy: https://en.cppreference.com/w/cpp/algorithm/copy

Milerius commented on src/order_manager/order.manager.hpp:
```diff
@@ -0,0 +1,52 @@
+        orders::orders_by_id orders_;
+        orders::executions_by_id executions_;
+
+        orders::executions_by_id find_new_executions(std::vector<orders::order> orders);
```
this function take by copy an std::vector, i think you want it by const reference or rvalue reference

Milerius commented on src/order_manager/order.manager.tests.cpp:
```diff
@@ -0,0 +1,71 @@
+
+namespace antara::mmbot::tests
+{
+    // TEST_CASE ("individual orders can be placed on the cex")
```
this file should be removed ?

DCPRevere replied:
I'm replacing the tests now.

Milerius replied:
Ok keep going !

Milerius commented on src/orders/orders.cpp:
```diff
@@ -18,12 +18,110 @@
+{
+    pair pair::of (std::string a, std::string b)
+    {
+        antara::pair pair = {{st_symbol{a}}, {st_symbol{b}}};
```
st_symbol{std::move(a)}, you take string by copy here, and you want to move the value inside the pair during the creation

Milerius commented on src/orders/orders.cpp:
```diff
@@ -18,12 +18,110 @@
+    // Order Builder
+
+    order_builder::order_builder(const st_order_id &id, const antara::pair &pair)
+    {
```
braces constructor initialization `: field_(value), field2_(value2)...`

DCPRevere committed:
fix: delete executions when their order finishes

DCPRevere committed:
style: use std::copy and transform

style: throw not_implemented

DCPRevere committed:
style: use braces list initialisation

DCPRevere committed:
style: use std::move

DCPRevere committed:
fix: add iterator include for windows

DCPRevere committed:
fix: move iterator include to the right file

Milerius committed:
style: local include instead of system one

Milerius committed:
fix: warn unused, and correct field initialization

Milerius committed:
fix: fix more warning/initialization

DCPRevere committed:
feat: pull live order polling into a separate function

Milerius committed:
feat(mock): add trompeloeil library (mocking)

DCPRevere committed:
tests: push broken testing code

Milerius committed:
feat(mock): fix compilation

DCPRevere committed:
feat: add tests to the OM

style: add cex and dex abstract classes
test: add cex and dex mocks

DCPRevere commented:
That was a big commit...

There are some outstanding tasks:
- the polling test has a bug, the test passes but shouldn't
- use a smart pointer for `dex` and `cex`
- doesn't compile on Windows because of strong string equality

Milerius committed:
fix: fix windows compilation

Milerius committed:
fix: fix windows compilation

Milerius committed:
improvements: code style and minor improvements

Milerius commented:
Need to fix strategy manager (use the function that is not used), and fix the unit tests too and commented code should be absolutely removed everywhere

Milerius commented on src/order_manager/order.manager.cpp:
```diff
@@ -0,0 +1,144 @@
+        for (const auto &id : order_ids) {
+            msg += id;
+        }
+        std::cout << msg << std::endl;
```
this debug should be removed

Milerius commented on src/order_manager/order.manager.cpp:
```diff
@@ -0,0 +1,144 @@
+#include <algorithm>
+#include <iterator>
+#include <loguru.hpp>
+#include <iostream>
```
`#include <iostream>` should be removed too once you remove debug msg

Milerius commented on src/order_manager/order.manager.cpp:
```diff
@@ -0,0 +1,144 @@
+        update_from_live();
+
+        auto order_ids = std::unordered_set<st_order_id>();
+        // std::transform(orders_.begin(), orders_.end(),
```
Dont forget to remove all commented code

Milerius commented on src/utils/mmbot_strong_types.hpp:
```diff
@@ -105,13 +112,30 @@ namespace std
     };
 
+    /*template<>
+    struct hash<antara::st_order_id>
```
This code should be removed too

Milerius commented on src/utils/exceptions.hpp:
```diff
@@ -0,0 +1,35 @@
+    {
+    public:
+        not_implemented() noexcept : std::runtime_error("Function has not been implemented.") {}
+        not_implemented(std::string s) noexcept : std::runtime_error("Function has not been implemented: " + s) {}
```
std::move(s) i think

DCPRevere committed:
style: remove commented code

DCPRevere committed:
fix: executions aren't mirrored multiple times

DCPRevere committed:
style: remove unused tests

DCPRevere committed:
style: move code into antara::mmbot namespace
